hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CalcLimitTransposeRule.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.opt.physical;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.rules.TransformationRule;
+import org.immutables.value.Value;
+
+import static java.util.Collections.singletonList;
+
+/**
+ * Logical rule that transposes {@link LimitPhysicalRel} with {@link CalcPhysicalRel}.
+ * <p>
+ * Before:
+ * <pre>
+ * CalcPhysicalRel[...]
+ *   LimitPhysicalRel[...]
+ * </pre>
+ * After:
+ * <pre>
+ * LimitPhysicalRel
+ *   CalcPhysicalRel[...]
+ * </pre>
+ */
+@Value.Enclosing
+public class CalcLimitTransposeRule extends RelRule<RelRule.Config> implements TransformationRule {
```

Comment:
The transformation here isn't to an equivalent rel. If you apply the limitation first, and then filter, it's not the same as if you filter first, and then apply the limitation.